sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateAntiJoin.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.plans.LeftAnti
+import org.apache.spark.sql.catalyst.plans.logical.{Join, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * This optimization rule detects and eliminate a LeftAntiJoin when buildSide is empty.
+ */
+case class EliminateAntiJoin(conf: SQLConf) extends Rule[LogicalPlan] {
+
+  private def canEliminate(plan: LogicalPlan): Boolean = plan match {
+    case LogicalQueryStage(_, stage: QueryStageExec) if stage.resultOption.get().isDefined
+      && stage.getRuntimeStatistics.rowCount.isDefined
+      && stage.getRuntimeStatistics.rowCount.get == 0L => true
+    case _ => false
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformDown {
+    // If the right side is empty, LeftAntiJoin simply returns the left side.
+    // Eliminate Join with left LogicalPlan instead.
+    case Join(left, right, LeftAnti, _, _) if canEliminate(right) =>
```

Comment:
This is not very useful, as the anti join is already very fast with the empty right side.

https://github.com/apache/spark/pull/29389/files#r467683621 is more useful, as for now we still need to consume all the left side data and return nothing. It will be a big improvement if we can directly convert the join node to an empty relation with AQE.

Reply:
> This is not very useful, as the anti join is already very fast with the empty right side.
> 
> https://github.com/apache/spark/pull/29389/files#r467683621 is more useful, as for now we still need to consume all the left side data and return nothing. It will be a big improvement if we can directly convert the join node to an empty relation with AQE.

In fact, we need to consume all the left side data and return all the rows when it's an EmptyHashedRelation.

Reply:
> In fact, we need to consume all the left side data and return all the rows when it's an EmptyHashedRelation.

Yes, but this rule doesn't change it. By returning the left side plan, we still need to consume all the data of the left side.